azurerm/resource_arm_application_gateway.go

```diff
@@ -0,0 +1,1685 @@
+		if config.ApplicationGatewayURLPathMapPropertiesFormat.DefaultBackendAddressPool != nil {
+			pathMap["default_backend_address_pool_name"] = path.Base(*config.ApplicationGatewayURLPathMapPropertiesFormat.DefaultBackendAddressPool.ID)
+			pathMap["default_backend_address_pool_id"] = *config.ApplicationGatewayURLPathMapPropertiesFormat.DefaultBackendAddressPool.ID
+		}
+
+		if config.ApplicationGatewayURLPathMapPropertiesFormat.DefaultBackendHTTPSettings != nil {
+			pathMap["default_backend_http_settings_name"] = path.Base(*config.ApplicationGatewayURLPathMapPropertiesFormat.DefaultBackendHTTPSettings.ID)
+			pathMap["default_backend_http_settings_id"] = *config.ApplicationGatewayURLPathMapPropertiesFormat.DefaultBackendHTTPSettings.ID
+		}
+
+		pathRules := make([]interface{}, 0, len(*config.ApplicationGatewayURLPathMapPropertiesFormat.PathRules))
+		for _, pathRuleConfig := range *config.ApplicationGatewayURLPathMapPropertiesFormat.PathRules {
+			rule := map[string]interface{}{
+				"id":   *pathRuleConfig.ID,
+				"name": *pathRuleConfig.Name,
+			}
+
+			if pathRuleConfig.ApplicationGatewayPathRulePropertiesFormat.BackendAddressPool != nil {
+				rule["backend_address_pool_name"] = path.Base(*pathRuleConfig.ApplicationGatewayPathRulePropertiesFormat.BackendAddressPool.ID)
+				rule["backend_address_pool_id"] = *pathRuleConfig.ApplicationGatewayPathRulePropertiesFormat.BackendAddressPool.ID
+			}
+
+			if pathRuleConfig.ApplicationGatewayPathRulePropertiesFormat.BackendHTTPSettings != nil {
+				rule["backend_http_settings_name"] = path.Base(*pathRuleConfig.ApplicationGatewayPathRulePropertiesFormat.BackendHTTPSettings.ID)
+				rule["backend_http_settings_id"] = *pathRuleConfig.ApplicationGatewayPathRulePropertiesFormat.BackendHTTPSettings.ID
+			}
+
+			paths := make([]interface{}, 0, len(*pathRuleConfig.ApplicationGatewayPathRulePropertiesFormat.Paths))
+			for _, rulePath := range *pathRuleConfig.ApplicationGatewayPathRulePropertiesFormat.Paths {
+				paths = append(paths, rulePath)
+			}
+			rule["paths"] = paths
+
+			pathRules = append(pathRules, rule)
+		}
+		pathMap["path_rule"] = pathRules
+
+		result = append(result, pathMap)
+	}
+
+	return result
+}
+
+func flattenApplicationGatewayAuthenticationCertificates(certs *[]network.ApplicationGatewayAuthenticationCertificate) []interface{} {
+	result := make([]interface{}, 0, len(*certs))
+
+	for _, config := range *certs {
+		certConfig := map[string]interface{}{
+			"id":   *config.ID,
+			"name": *config.Name,
+		}
+
+		result = append(result, certConfig)
+	}
+
+	return result
+}
+
+func flattenApplicationGatewaySslCertificates(certs *[]network.ApplicationGatewaySslCertificate) []interface{} {
+	result := make([]interface{}, 0, len(*certs))
+
+	for _, config := range *certs {
+		certConfig := map[string]interface{}{
+			"id":               *config.ID,
+			"name":             *config.Name,
+			"public_cert_data": *config.ApplicationGatewaySslCertificatePropertiesFormat.PublicCertData,
```

Comment:
do we need to add a check for `if config.ApplicationGatewaySslCertificatePropertiesFormat != nil` here?